code/modules/mob/living/simple_animal/hostile/megafauna/adamant.dm

```diff
@@ -0,0 +1,278 @@
+	icon = 'icons/mob/lavaland/96x96megafauna.dmi'
+	faction = list("mining")
+	weather_immunities = list("lava","ash")
+	speak_emote = list("bellows")
+	density = 1
+	armour_penetration = 0
+	melee_damage_lower = 60
+	melee_damage_upper = 60
+	speed = 1
+	move_to_delay = 10
+	wander = 0
+	ranged = 1
+	flying = 1
+	mob_size = MOB_SIZE_LARGE
+	pixel_x = -32
+	aggro_vision_range = 18
+	idle_vision_range = 5
+	del_on_death = 0
+	medal_type = MEDAL_PREFIX
+	score_type = ADAMANT_SCORE
+
+	deathmessage = "collapses into a heap of twisted metal."
+	death_sound = 'sound/magic/demon_dies.ogg'
+	damage_coeff = list(BRUTE = 0, BURN = 0, TOX = 0, CLONE = 0, STAMINA = 0, OXY = 0)
+
+	var/state = STATE_NORMAL
+	var/obj/item/device/gps/internal
+	var/struck = 0
+
+
+/obj/item/device/gps/internal/colossus
+	icon_state = null
+	gpstag = "Hardened Signal"
+	desc = "It's a signal, how can it be hard?"
+	invisibility = 100
+
+/mob/living/simple_animal/hostile/megafauna/adamant/AttackingTarget()
+	if(state == STATE_FREEZE)
+		return 0
+	..()
+
+
+/mob/living/simple_animal/hostile/megafauna/adamant/OpenFire()
+	if(state)
+		return
+
+	state = STATE_ATTACKING
+	if(get_dist(src, target) < 7)
+		if(prob(70))
+			projectile_burst()
+		else
+			bolt_volley()
+		state = STATE_NORMAL
+	else
+		telefrag(target)
+
+
+/mob/living/simple_animal/hostile/megafauna/adamant/proc/shoot_projectile(turf/marker)
+	if(!marker)
+		return
+	var/turf/startloc = get_turf(src)
+	var/obj/item/projectile/P = new /obj/item/projectile/adamant(startloc)
+	P.current = startloc
+	P.starting = startloc
+	P.firer = src
+	P.yo = marker.y - startloc.y
+	P.xo = marker.x - startloc.x
+	if(target)
+		P.original = target
+	else
+		P.original = marker
+	P.fire()
+
+
+/obj/item/projectile/adamant
+	name ="energy bolt"
+	icon_state= "chronobolt"
+	damage = 25
+	armour_penetration = 100
+	speed = 2
+	eyeblur = 0
+	damage_type = BURN
+	pass_flags = PASSTABLE
+
+
+/obj/item/projectile/adamant/on_hit(atom/target, blocked = 0)
+	..()
+	explosion(target, 0, 1, 2)
+	return 1
+
+
+/mob/living/simple_animal/hostile/megafauna/adamant/Move()
+	if(state == STATE_FREEZE)
+		return 0
+	if(!stat)
+		playsound(src.loc, 'sound/effects/meteorimpact.ogg', 200, 1)
+	..()
+
+
+/mob/living/simple_animal/hostile/megafauna/adamant/devour(mob/living/L)
+	visible_message(
+		"<span class='danger'>[src] burns [L] to ash!</span>",
+		"<span class='userdanger'>You burn [L] to ash, restoring your health!</span>")
+	adjustBruteLoss(-L.maxHealth/2)
+	L.dust()
+
+
+/mob/living/simple_animal/hostile/megafauna/adamant/proc/projectile_burst()
+	playsound(get_turf(src), 'sound/magic/clockwork/invoke_general.ogg', 200, 1, 2)
+	for(var/turf/T in range(2, target))
+		if(prob(30))
+			shoot_projectile(T)
+	sleep(20)
+
+
+/mob/living/simple_animal/hostile/megafauna/adamant/proc/telefrag(mob/living/victim)
+	var/matrix/ntransform = matrix(transform)
+	animate(src, transform=ntransform.Scale(10), alpha=0, 10)
+	density = 0
+	sleep(50)
+	freeze()
```

Comment:
The rest of this proc will wait for the freeze to finish, presumably you don't want that and it should be below the for loop.